engine/api/lastupdate.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/engine/api/businesscontext"
+	"github.com/ovh/cds/engine/api/cache"
+	"github.com/ovh/cds/engine/api/sessionstore"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+)
+
+// LastUpdateBrokerSubscribe is the information need to to subscribe
+type LastUpdateBrokerSubscribe struct {
+	UIID  string
+	User  *sdk.User
+	Queue chan string
+}
+
+// LastUpdateBroker keeps connected client of the current route,
+type LastUpdateBroker struct {
+	clients    map[string]*LastUpdateBrokerSubscribe
+	newClients chan *LastUpdateBrokerSubscribe
+	messages   chan string
+}
+
+var lastUpdateBroker *LastUpdateBroker
+
+func InitLastUpdateBroker(c context.Context, DBFunc func() *gorp.DbMap) {
+	lastUpdateBroker = &LastUpdateBroker{
+		make(map[string]*LastUpdateBrokerSubscribe),
+		make(chan *LastUpdateBrokerSubscribe),
+		make(chan string),
+	}
+
+	// Start cache Subscription
+	go CacheSubscribe(c, lastUpdateBroker.messages)
+
+	// Start processing events
+	go lastUpdateBroker.Start(c, DBFunc())
+}
+
+// CacheSubscribe subscribe to a channel and push received message in a channel
+func CacheSubscribe(c context.Context, cacheMsgChan chan<- string) {
+	pubSub := cache.Subscribe("lastUpdates")
+	tick := time.NewTicker(250 * time.Millisecond).C
+	for {
+		select {
+		case <-c.Done():
+			if c.Err() != nil {
+				log.Warning("lastUpdate.CacheSubscribe> Exiting: %v", c.Err())
+				return
+			}
+		case <-tick:
+			msg, err := cache.GetMessageFromSubscription(pubSub, c)
+			if err != nil {
+				log.Warning("lastUpdate.CacheSubscribe> Cannot get message %s: %s", msg, err)
+				time.Sleep(5 * time.Second)
+				continue
+			}
+			cacheMsgChan <- msg
+		}
+	}
+}
+
+// Start the broker
+func (b *LastUpdateBroker) Start(c context.Context, db gorp.SqlExecutor) {
```

Comment:
this will exit if we lost database connection. You should get the db in the for loop